drivers/counter: stm32: Convert configuration from Kconfig to device tree

gautierg-st opened the pull request:
Use device tree to configure rtc counter driver instead of Kconfig for:
- RTC source clock selection (thanks to use of dts domain clock)
- LSE drive capability (use lse dts property)

gautierg-st committed:
drivers: counter: Add support for STM32 RTC source from dts

Add support for STM32 RTC source from dts

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
include: zephyr: dt-bindings: clock: Add RTC clk src defines for STM32

For all STM32 series, add some defines to select the RTC clock source.
This change is necessary to be able to select the RTC clock source in
dts instead of Kconfig

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
boards: arm: Add RTC clock source for STM32 boards dts

For every board using an STM32, add the RTC clock source in its dts

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

erwango commented on include/zephyr/dt-bindings/clock/stm32f0_clock.h:
```diff
@@ -27,6 +27,8 @@
 /** PLL clock */
 #define STM32_SRC_PLLCLK	0x006
 
+#define STM32_SRC_LSI		0x007
```
Please move to `/** Fixed clocks  */` with LSE/HSI.
Apply everywhere applicable.

gautierg-st replied:
Done

erwango commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -406,55 +430,55 @@ static int rtc_stm32_init(const struct device *dev)
+		/* Configure driving capability */
+		LL_RCC_LSE_SetDriveCapability(STM32_LSE_DRIVING << RCC_BDCR_LSEDRV_Pos);
+
+#endif		/*
```
Use one space instead of tab here.

erwango commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -406,55 +430,55 @@ static int rtc_stm32_init(const struct device *dev)
-	while (LL_RCC_LSI1_IsReady() != 1) {
-	}
+		LL_RCC_LSI1_Enable();
+		while (LL_RCC_LSI1_IsReady() != 1) {
+		}
 #else
-	LL_RCC_LSI_Enable();
-	while (LL_RCC_LSI_IsReady() != 1) {
-	}
+		LL_RCC_LSI_Enable();
+		while (LL_RCC_LSI_IsReady() != 1) {
+		}
```
Clocks activation should now be performed y clock_control driver and hence removed from here.

erwango commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -406,55 +430,55 @@ static int rtc_stm32_init(const struct device *dev)
+		!defined(CONFIG_SOC_SERIES_STM32F1X) && \
+		!defined(CONFIG_SOC_SERIES_STM32L1X)
+
+		/* Configure driving capability */
+		LL_RCC_LSE_SetDriveCapability(STM32_LSE_DRIVING << RCC_BDCR_LSEDRV_Pos);
```
This should also be done by clock_control driver

erwango commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -406,55 +430,55 @@ static int rtc_stm32_init(const struct device *dev)
-	* && !CONFIG_SOC_SERIES_STM32F1X
-	* && !CONFIG_SOC_SERIES_STM32L1X
-	*/
+		!defined(CONFIG_SOC_SERIES_STM32F2X) && \
+		!defined(CONFIG_SOC_SERIES_STM32F1X) && \
+		!defined(CONFIG_SOC_SERIES_STM32L1X)
+
+		/* Configure driving capability */
+		LL_RCC_LSE_SetDriveCapability(STM32_LSE_DRIVING << RCC_BDCR_LSEDRV_Pos);
+
+#endif		/*
+		 * !CONFIG_SOC_SERIES_STM32F4X
+		 * && !CONFIG_SOC_SERIES_STM32F2X
+		 * && !CONFIG_SOC_SERIES_STM32F1X
+		 * && !CONFIG_SOC_SERIES_STM32L1X
+		 */
 
 #if defined(CONFIG_COUNTER_RTC_STM32_LSE_BYPASS)
-	LL_RCC_LSE_EnableBypass();
+		LL_RCC_LSE_EnableBypass();
 #endif /* CONFIG_COUNTER_RTC_STM32_LSE_BYPASS */
 
-	LL_RCC_LSE_Enable();
+		LL_RCC_LSE_Enable();
 
-	/* Wait until LSE is ready */
-	while (LL_RCC_LSE_IsReady() != 1) {
-	}
+		/* Wait until LSE is ready */
+		while (LL_RCC_LSE_IsReady() != 1) {
+		}
 
 #if STM32_MSI_PLL_MODE
-	/* Enable MSI hardware auto calibration */
-	LL_RCC_MSI_EnablePLLMode();
+		/* Enable MSI hardware auto calibration */
+		LL_RCC_MSI_EnablePLLMode();
 #endif
```
Ditto

erwango commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -57,9 +57,23 @@ LOG_MODULE_REGISTER(counter_rtc_stm32, CONFIG_COUNTER_LOG_LEVEL);
+
 struct rtc_stm32_config {
 	struct counter_config_info counter_info;
-	struct stm32_pclken pclken;
```
Why removing clock_config from this structure ?

gautierg-st committed:
dts: bindings: clock: add bypass property to stm32 lse

For STM32 LSE clock, add LSE bypass property (defaulting to false)

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
include: zephyr: drivers: clock_control: Add stm32 lse bypass

Add a define to get the value of the LSE bypass property from dts

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: clock_control: clock_stm32: Add support for lse bypass

Add support for LSE bypass for all STM32 series

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

mbolivar-nordic reviewed (commented):
ACK on the bindings changes

erwango commented on include/zephyr/dt-bindings/clock/stm32f4_clock.h:
```diff
@@ -26,5 +26,43 @@
 #define STM32_SRC_PLL_P	0x001
 #define STM32_SRC_PLL_Q	0x002
 #define STM32_SRC_PLL_R	0x003
+#define STM32_SRC_LSE	0x004
+#define STM32_SRC_LSI	0x005
```
/** Fixed clocks */

erwango commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -392,69 +392,19 @@ static int rtc_stm32_init(const struct device *dev)
-
-#endif /* CONFIG_COUNTER_RTC_STM32_CLOCK_SRC */
+	/* Enable RTC clock source */
+	clock_control_configure(clk, (clock_control_subsys_t *) &cfg->pclken[1],
```
Add error checking.

erwango commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -494,40 +444,52 @@ static int rtc_stm32_init(const struct device *dev)
 static struct rtc_stm32_data rtc_data;
 
+#if DT_INST_NUM_CLOCKS(0) == 1
+static const struct stm32_pclken rtc_clk[] = {
+	STM32_CLOCK_INFO(0, DT_DRV_INST(0)),
+	/* Use Kconfig to configure source clocks fields */
+	/* Fortunately, values are consistent across enabled series */
+#ifdef COUNTER_RTC_STM32_CLOCK_LSI
+	{.bus = STM32_SRC_LSI, .enr = RTC_SEL(2)}
+#else
+	{.bus = STM32_SRC_LSE, .enr = RTC_SEL(1)}
+#endif
+};
```
Add a warning in this case which is now deprecated.

erwango commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -392,69 +392,19 @@ static int rtc_stm32_init(const struct device *dev)
-
 	LL_PWR_EnableBkUpAccess();
 
-#if defined(CONFIG_COUNTER_RTC_STM32_BACKUP_DOMAIN_RESET)
```
Kconfig symbol should be cleaned up.

erwango reviewed (commented):
Some comments remaining otherwise looks good.
Additionally, please update release notes (3.3) with deprecated and removed Kconfig symbols

gautierg-st committed:
drivers: counter: Remove unused Kconfig symbols for stm32 rtc

Remove a set of Kconfig symbols for configuring the RTC on STM32
that are no longer useful

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

erwango commented on doc/releases/release-notes-3.3.rst:
```diff
@@ -189,6 +206,14 @@ USB
+
+    * STM32 SoCs:
+
+      * LSE Clock binding: new ``lse-bypass`` property
```
```suggestion
      * :dtcompatible: `st,stm32-lse-clock`: new ``lse-bypass`` property
```

gautierg-st committed:
doc: releases: Update release notes with stm32 rtc modification

Update release notes 3.3 with the different modifications brought
to the STM32 RTC counter including new functionality, deprecated
and removed Kconfig

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

FRASTM commented on boards/arm/nucleo_l152re/nucleo_l152re.dts:
```diff
@@ -85,6 +89,8 @@
 };
 
 &rtc {
+	clocks = <&rcc STM32_CLOCK_BUS_APB1 0x10000000>,
```
Check RM0394 for the stm32l41x/42x/43x/44x/45x/46x  : clocks = <&rcc STM32_CLOCK_BUS_APB1 **0x400**>,

FRASTM reviewed (commented):
Actually in several stm32 devices, the RTC peripheral has a Enable bit in the RCC BDCR register plus a RTC  clock  source selection on RTCSEL bits where default value 00 is "no clock" without RCC bus clock control bit.
(ex on the stm32G4 serie).

Here, the RTC node in the DTS is configured with  RTC SEL is LSI. Consequently the LSI must also be enabled.
--> That will now make the RTC running on each stm32 platform (and probably increase the power consumption).
